metals/src/main/scala/scala/meta/internal/metals/LSPFileSystemProvider.scala

```diff
@@ -0,0 +1,249 @@
+package scala.meta.internal.metals
+
+import java.nio.charset.StandardCharsets
+import java.nio.file.Files
+import java.nio.file.Path
+import java.util.stream.Collectors
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.util.control.NonFatal
+
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.metals.clients.language.MetalsLanguageClient
+
+final case class FSReadDirectoryResponse(
+    name: String,
+    isFile: Boolean,
+)
+final case class FSReadDirectoriesResponse(
+    name: String,
+    directories: Array[FSReadDirectoryResponse],
+    error: String,
+)
+
+final case class FSReadFileResponse(
+    name: String,
+    value: String,
+    error: String,
+)
+
+final case class FSStatResponse(
+    name: String,
+    isFile: Boolean,
+    error: String,
+)
+
+final class LSPFileSystemProvider(
```

Comment:
I wonder if creating out own FileSystemProvider wouldn't actually cause less work (less changes and no need for a mapper), we could handle those URIs like normal ones and paths should be fine also. We would only mos likely need to map them for DAP

A lot of the methods in https://docs.oracle.com/javase/8/docs/api/java/nio/file/spi/FileSystemProvider.html can be left blank.

I can try and take a look at it if you don't have the time (unless you think it's not a great idea)

Reply:
@tgodzik I have created a working FileSystemProvider in #3750.  It's pretty much complete - I just never pushed the last version because it was slow.  I guess I could spend some time optimising.  My other issue was the URL translation for the custom URLConnection class needed - it's an area that I don't know much about so I'm not sure how many types of URL have to be dealt with for the URLClassLoader to work with the custom FileSystem - I ended up going round in circles a bit there.  You're right in that it's less invasive to the rest of the codebase - no mapping required except for the debugger.  If you want, I can push the latest version of #3750 and you can decide for yourself if you think it's a better approach.

Reply:
If the issue with `URLClassLoader` is complicated then maybe the current approach would work better. Let me think about it for a bit.